File: src/lib/scrapers/amazon.ts
import * as cheerio from "cheerio";
import type { PriceResult } from "../types";

const BASE = "https://www.amazon.nl";

export async function searchAmazon(query: string): Promise<PriceResult[]> {
  const url = `${BASE}/s?k=${encodeURIComponent(query)}&ref=nb_sb_noss`;

  const res = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36",
      "Accept-Language": "nl-NL,nl;q=0.9",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    },
    next: { revalidate: 300 },
  });

  if (!res.ok) throw new Error(`Amazon HTTP ${res.status}`);

  const html = await res.text();
  const $ = cheerio.load(html);
  const results: PriceResult[] = [];

  $('[data-component-type="s-search-result"]').each((_, el) => {
    // Title
    const name = $(el).find("h2 .a-text-normal, h2 span").first().text().trim();

    // Price — the screen-reader span has the cleanest value, e.g. "€ 549,99"
    const priceText = $(el)
      .find(".a-price .a-offscreen")
      .first()
      .text()
      .replace(/[^\d,]/g, "")
      .replace(",", ".");
    const price = parseFloat(priceText);

    // Link — relative ASIN path
    const href = $(el).find("h2 a.a-link-normal").attr("href") ?? "";
    const link = href.startsWith("http") ? href : `${BASE}${href}`;

    // Image
    const img = $(el).find(".s-image").attr("src");

    if (name && !isNaN(price) && price > 0) {
      results.push({
        retailer: "amazon",
        name,
        priceEur: price,
        url: link,
        imageUrl: img,
        inStock: true, // if a price is shown, it's available
      });
    }
  });

  return results.slice(0, 10);
}

File: scrapers/corebuild_scrapers/retailers/amazon.py
"""Amazon.nl: werkt vanaf residentiële IP's; datacenter-IP's worden geblokkeerd.

Zet AMAZON_ASSOCIATE_TAG in .env.local om affiliate-tags aan links toe te voegen.
"""

import os
from urllib.parse import quote

from bs4 import BeautifulSoup

from .common import MAX_RESULTS, get, parse_nl_price

BASE = "https://www.amazon.nl"


def search(query: str) -> list[dict]:
    res = get(f"{BASE}/s?k={quote(query)}&ref=nb_sb_noss")
    soup = BeautifulSoup(res.text, "html.parser")
    tag = os.environ.get("AMAZON_ASSOCIATE_TAG")
    results = []

    for el in soup.select('[data-component-type="s-search-result"]'):
        title = el.select_one("h2 .a-text-normal") or el.select_one("h2 span")
        price_el = el.select_one(".a-price .a-offscreen")
        img = el.select_one(".s-image")
        asin = el.get("data-asin")

        if not title or not price_el:
            continue
        price = parse_nl_price(price_el.get_text())
        if price is None:
            continue

        if asin:
            url = f"{BASE}/dp/{asin}"
        else:
            link = el.select_one('a.a-link-normal[href*="/dp/"]') or el.select_one("h2 a")
            href = link.get("href", "") if link else ""
            url = href if href.startswith("http") else f"{BASE}{href}"
        if tag:
            url += f"?tag={tag}"

        results.append(
            {
                "retailer": "amazon",
                "name": title.get_text(strip=True),
                "price_eur": price,
                "url": url,
                "image_url": img.get("src") if img else None,
                "in_stock": True,  # prijs getoond = leverbaar
            }
        )
        if len(results) >= MAX_RESULTS:
            break

    return results
